myfitnesspal/client.py

```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+            except Exception:
+                logger.warning(f"Unable to submit food to MyFitnessPal: {error}")
+
+        elif not result.ok:
+            logger.warning(
+                f"Request Error - Unable to submit food to MyFitnessPal: status code: {result.status_code}"
+            )
+            return None
+
+    def set_new_goal(
+        self,
+        energy: float = None,
+        energy_unit: str = None,
+        carbohydrates: float = None,
+        protein: float = None,
+        fat: float = None,
+        percent_carbohydrates: float = None,
+        percent_protein: float = None,
+        percent_fat: float = None,
+        # TODO Add micro nutritions
+        # saturated_fat: float = None,
+        # polyunsaturated_fat: float = None,
+        # monounsaturated_fat: float = None,
+        # trans_fat: float = None,
+        # fiber: float = None,
+        # sugar: float = None,
+        # cholesterol: float = None,
+        # sodium: float = None,
+        # potassium: float = None,
+        # vitamin_a: float = None,
+        # vitamin_c: float = None,
+        # calcium: float = None,
+        # iron: float = None,
+        assign_exercise_energy="nutrient_goal",
+    ):
+        """Function to update your nutrition goals. Function will return True if successful."""
+        # FROM MFP JS:
+        # var calculated_energy = 4 * parseFloat(this.get('carbGrams')) + 4 * parseFloat(this.get('proteinGrams')) + 9 * parseFloat(this.get('fatsGrams'));
+
+        # Get User Default Unit Preference
+        if energy_unit != "calories" and energy_unit != "kilojoules":
+            energy_unit = self.user_metadata["unit_preferences"]["energy"]
+
+        # Get authenticity token and current values
+        url = parse.urljoin(self.BASE_URL_SECURE, "account/my_goals")
+        today = datetime.datetime.now().strftime("%Y-%m-%d")
+
+        """
+        Will leave this here as reminder why we do not use self._get_json_for_url() at the moment.
+        # No Way implemented to send token by calling self._get_json_for_url
+        old_goals_url = parse.urljoin(self.BASE_API_URL, f"v2/nutrient-goals?date={today}")
+        old_goals = self._get_json_for_url(old_goals_url)
+        """
+
+        # Build header for API-requests
+        auth_header = self.session.headers
+        auth_header["authorization"] = f"Bearer {self.access_token}"
+        auth_header["mfp-client-id"] = "mfp-main-js"
+        auth_header["mfp-user-id"] = f"{self.user_id}"
+
+        # Get Request for old goal values
+        old_goals_url = parse.urljoin(
+            self.BASE_API_URL, f"v2/nutrient-goals?date={today}"
+        )
+        old_goals_document = self.session.get(old_goals_url, headers=auth_header)
+        old_goals = json.loads(old_goals_document.text)
+
+        # Marcro Calculation
+        # If no macro goals were provided calculate them with percentage value
+        if carbohydrates is None or protein is None or fat is None:
```

Comment:
I'm sort of of the mind right now that maybe we shouldn't try to handle missing information at all; it'd make the logic a lot simpler below, and I wonder what use case we're imagining for somebody to only provide a subset of these?

Reply:
The idea was that you can pass absolut values or percentage values. I think there are use cases for both.
But i am with you that the logic looks really complex. Do you think it would be better to split it into two functions?

Reply:
I added an explaining doc string to the function. How it works should be clear now.